src/node.js

```diff
@@ -390,6 +391,78 @@
     }
   };
 
+  startup.processPromises = function() {
+    // this should be kept in sync with PromiseRejectEvent in v8.h
+    var kPromiseRejectWithNoHandler = 0;
+    var kPromiseHandlerAddedAfterReject = 1;
+
+    var timers = NativeModule.require('timers');
+    var unhandledPromises = new WeakMap();
+    // A hack to work around uniterable WeakMaps.
+    var unhandledPromiseKeys = [];
+    var fullTurnQueued = false;
+
+    function queueFullTurn() {
+      if (!fullTurnQueued) {
+        fullTurnQueued = true;
+        timers.setImmediate(function() {
+          process.nextTick(afterFullTurn);
```

Comment:
What I saw in @domenic's proposal is that this should be done in the end of the current event loop turn. `setImmediate` doesn't give you that.